File: ai/src/main/java/org/pingel/bayes/Case.java
package org.pingel.bayes;

import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.Vector;

public class Case implements Comparable<Case>
{
	private Map<RandomVariable,Value> assignments = new TreeMap<RandomVariable, Value>();

	public Case() {}

	public Case(RandomVariable var, Value val)
	{
	    assign(var, val);
	}

	public Set<RandomVariable> getVariables()
	{
	    return assignments.keySet();
	}
	
	public int size()
	{
		return assignments.keySet().size();
	}

	public Value valueOf(RandomVariable var) {
		return assignments.get(var);
	}

	public List<Value> valuesOf(List<RandomVariable> vars) {

	    // Note: this may contain null entries if assignments.keySet()
	    // is a strict subset of vars
	    
	    List<Value> result = new Vector<Value>();
	    for(int i=0; i < vars.size(); i++) {
	        result.add(assignments.get(vars.get(i)));
	    }
	    return result;
	}
	
	public void assign(RandomVariable rv, Value val)
	{
		assignments.put(rv, val);
	}

	public void assign(List<RandomVariable> vars, List<Value> vals)
	{
	    for(int i=0; i < vars.size(); i++ ) {
	        RandomVariable var = vars.get(i);
	        Value val = vals.get(i);
			assignments.put(var, val);
	    }
	}

	public boolean isSupersetOf(Case other)
	{
		Iterator<RandomVariable> it = other.assignments.keySet().iterator();
		while( it.hasNext() ) {
		    RandomVariable var = it.next();
		    Value otherVal = other.valueOf(var);
		    Value thisVal = valueOf(var);
			if ( otherVal != null &&
				 thisVal != null &&
				 ! thisVal.equals(otherVal) ) {
				return false;
			}
		}
		return true;
	}
	
	public Case copy()
	{
		Case result = new Case();
        result.assignments = new TreeMap<RandomVariable, Value>();
        result.assignments.putAll(assignments);
		return result;
	}

	public Case projectToVars(List<RandomVariable> pVars)
	{
		Case result = new Case();

		for( RandomVariable var : pVars ) {
			result.assign(var, valueOf(var));
		}
		
		return result;
	}

	public boolean equals(Object o) {
        if( o instanceof Case ) {
            return compareTo((Case)o) == 0;
        }
        else {
            return false;
        }
	}
	
	public int compareTo(Case other)
	{
	    if( assignments.size() < other.assignments.size() ) {
			return -1;
		}
		if( assignments.size() > other.assignments.size() ) {
			return 1;
		}
		
		for( RandomVariable var : assignments.keySet() ) {
			Value myValue = assignments.get(var);
			Value otherValue = other.assignments.get(var);
			
			if( ! myValue.equals(otherValue) ) {
				return myValue.compareTo(otherValue);
			}
		}

		return 0;
	}
	
    public String toString()
    {
        String result = "";
        for( RandomVariable rv : assignments.keySet() ) {
            // System.out.println("rv = " + rv.name);
            result += rv.name + " = ";
            if(  assignments.get(rv) == null ) {
                result += "null";
            }
            else {
                result += assignments.get(rv).toString();
            }
            result += ", ";
        }
        return result;
    }
    
    public String toOrderedString(RandomVariable... rvs)
    {
    		String result = "";
    	
    		for(int i=0; i < rvs.length; i++ ) {
    			result += rvs[i].name + " = " + assignments.get(rvs[i]);
    			if( i < rvs.length - 1 ) {
    				result += ", ";
    			}
    		}
    		
    		return result;
    }
    
    public String toOrderedString(List<RandomVariable> vs)
    {
    		String result = "";
		
		Iterator<RandomVariable> it = vs.iterator();
		for(int i=0; i < vs.size(); i++ ) {
			RandomVariable var = it.next();
			result += var.name + " = " + assignments.get(var);
			if( it.hasNext() ) {
				result += ", ";
			}
		}
		
		return result;
	}
}

File: ai/src/main/scala/org/pingel/bayes/Case.scala
package org.pingel.bayes

import org.pingel.gestalt.core.Value

// OLD VERSION HAD THESE DATA MEMBERS IN CONSTRUCTOR:
// rv: RandomVariable, value: Value

case class Case() extends Comparable[Case] {
  var assignments = Map[RandomVariable, Value]() // NOTE: was TreeMap

  // OLD VERSION DID THIS: assign(rv, value)

  def getVariables(): Set[RandomVariable] = assignments.keySet

  def size() = assignments.keySet.size

  def valueOf(variable: RandomVariable): Value = {
    assignments(variable)
  }

  def valuesOf(vars: List[RandomVariable]): List[Value] = {
    // Note: this may contain null entries if assignments.keySet()
    // is a strict subset of vars
    vars map { assignments(_) }
  }

  def assign(rv: RandomVariable, value: Value): Unit = {
    assignments += rv -> value
  }

  def assign(vars: List[RandomVariable], vals: List[Value]): Unit = {
    for (i <- 0 until vars.size) {
      assignments += vars(i) -> vals(i)
    }
  }

  def isSupersetOf(other: Case): Boolean = {
    val it = other.assignments.keySet.iterator
    while (it.hasNext) {
      val variable = it.next
      val otherVal = other.valueOf(variable)
      val thisVal = valueOf(variable)
      if (otherVal != null && thisVal != null && !thisVal.equals(otherVal)) {
        return false
      }
    }
    true
  }

  def copy(): Case = {
    var result = new Case()
    for ((rv, value) <- assignments) {
      result.assignments += rv -> value
    }
    result
  }

  def projectToVars(pVars: List[RandomVariable]): Case = {
    var result = new Case()
    for (variable <- pVars) {
      result.assign(variable, valueOf(variable))
    }
    result
  }

  override def equals(o: Object) = o match {
    case c: Case => compareTo(c) == 0
    case _ => false
  }

  def compareTo(other: Case): Integer = {
    if (assignments.size < other.assignments.size) {
      return -1
    }
    if (assignments.size > other.assignments.size) {
      return 1
    }
    for (variable <- assignments.keySet) {
      val myValue = assignments(variable)
      val otherValue = other.assignments(variable)
      if (!myValue.equals(otherValue)) {
        return myValue.compareTo(otherValue)
      }
    }
    0
  }

  override def toString(): String = {
    var result = ""
    for (rv <- assignments.keySet) {
      // System.out.println("rv = " + rv.name);
      result += rv.name + " = ";
      if (assignments.contains(rv)) {
        result += assignments(rv).toString()
      } else {
        result += "null"
      }
      result += ", "
    }
    result
  }

  def toOrderedString(vs: List[RandomVariable]): String =
    (for (variable <- vs) yield variable.getName + " = " + assignments(variable)).mkString(", ")

}
